google-youtube-upload.html

```diff
@@ -21,447 +30,687 @@
+     *
+     * @method uploadFile
+     * @param {object} file File object corresponding to the video to upload.
+     */
+    uploadFile: function(file) {
+      var metadata = {
+        snippet: {
+          title: this.videoTitle,
+          description: this.description,
+          tags: this.tags,
+          categoryId: this.categoryId
         },
-        _selectedFile: {
-          type: Object,
-          value: null
+        status: {
+          privacyStatus: this.privacyStatus
         }
-      },
-
-      _uploadStartTime: 0,
+      };
+
+      var uploader = new MediaUploader({
+        baseUrl: 'https://www.googleapis.com/upload/youtube/v3/videos',
+        file: file,
+        token: this._accessToken,
+        metadata: metadata,
+        params: {
+          part: Object.keys(metadata).join(',')
+        },
 
-      ready: function() {
-        document.addEventListener('google-signin-success', function(e) {
+        onError: function(data) {
+          var message = data;
+
+          // Assuming the error is raised by the YouTube API, data will be
+          // a JSON string with error.message set. I am not 100% sure that's the
+          // only time onError will be raised, though.
+          try {
+            var errorResponse = JSON.parse(data);
+            message = errorResponse.error.message;
+          } finally {
+            this.fire('youtube-upload-fail', message);
+          }
+        }.bind(this),
+
+        onProgress: function(data) {
+          var currentTime = Date.now();
+          var bytesUploaded = data.loaded;
+          var totalBytes = data.total;
+
+          // The times are in millis, so we need to divide by 1000 to get seconds.
+          var bytesPerSecond = bytesUploaded / ((currentTime - this._uploadStartTime) / 1000);
+          var estimatedSecondsRemaining = (totalBytes - bytesUploaded) / bytesPerSecond;
+          var fractionComplete = bytesUploaded / totalBytes;
+
+          this.fire('youtube-upload-progress', {
+            progressEvent: data,
+            bytesPerSecond: bytesPerSecond,
+            estimatedSecondsRemaining: estimatedSecondsRemaining,
+            fractionComplete: fractionComplete
+          });
+        }.bind(this),
+
+        onComplete: function(data) {
+          var uploadResponse = JSON.parse(data);
+          // Need to use _setVideoId() because videoId is read-only.
+          this._setVideoId(uploadResponse.id);
+          this.fire('youtube-upload-complete', uploadResponse.id);
+
+          this._pollForVideoStatus();
+        }.bind(this)
+      });
+
+      this.fire('youtube-upload-start', file);
+      // This won't correspond to the *exact* start of the upload, but it should be close enough.
+      this._uploadStartTime = Date.now();
+      uploader.upload();
+
+      this._transitionToThanks();
+    },
+
+    // Private helper methods.
+    _initializeMediaRecorder: function() {
+      if (this._mediaRecorder) {
+        return Promise.resolve();
```

Comment:
Given that the MediaRecorder API needs to be implemented in order for this line to execute, I don't think there's a real risk of it running on a browser that doesn't support Promises.